x/icqoracle/keeper/icq.go

```diff
@@ -0,0 +1,189 @@
+	c.callbacks[id] = fn.(ICQCallback)
+	return c
+}
+
+func (c ICQCallbacks) RegisterICQCallbacks() icqtypes.QueryCallbacks {
+	return c.
+		AddICQCallback(ICQCallbackID_OsmosisClPool, ICQCallback(OsmosisClPoolCallback))
+}
+
+// Submits an ICQ to get a concentrated liquidity pool from Osmosis' store
+func (k Keeper) SubmitOsmosisClPoolICQ(
+	ctx sdk.Context,
+	tokenPrice types.TokenPrice,
+) error {
+	k.Logger(ctx).Info(utils.LogWithPriceToken(tokenPrice, "Submitting OsmosisClPool ICQ"))
+
+	params := k.GetParams(ctx)
+
+	osmosisPoolId, err := strconv.ParseUint(tokenPrice.OsmosisPoolId, 10, 64)
+	if err != nil {
+		k.Logger(ctx).Error(utils.LogWithPriceToken(tokenPrice, "Error converting osmosis pool id '%s' to uint64, error '%s'", tokenPrice.OsmosisPoolId, err.Error()))
+		return err
+	}
+
+	tokenPriceBz, err := k.cdc.Marshal(&tokenPrice)
+	if err != nil {
+		k.Logger(ctx).Error(utils.LogWithPriceToken(tokenPrice, "Error serializing tokenPrice '%+v' to bytes, error '%s'", tokenPrice, err.Error()))
+		return err
+	}
+
+	queryId := fmt.Sprintf("%s|%s|%s|%d", tokenPrice.BaseDenom, tokenPrice.QuoteDenom, tokenPrice.OsmosisPoolId, ctx.BlockHeight())
+	query := icqtypes.Query{
+		Id:              queryId,
+		ChainId:         params.OsmosisChainId,
+		ConnectionId:    params.OsmosisConnectionId,
+		QueryType:       icqtypes.CONCENTRATEDLIQUIDITY_STORE_QUERY_WITH_PROOF,
+		RequestData:     icqtypes.FormatOsmosisKeyPool(osmosisPoolId),
+		CallbackModule:  types.ModuleName,
+		CallbackId:      queryId,
+		CallbackData:    tokenPriceBz,
+		TimeoutDuration: time.Duration(params.IcqTimeoutSec) * time.Second,
+		TimeoutPolicy:   icqtypes.TimeoutPolicy_RETRY_QUERY_REQUEST,
+	}
+	if err := k.icqKeeper.SubmitICQRequest(ctx, query, true); err != nil {
+		k.Logger(ctx).Error(utils.LogWithPriceToken(tokenPrice, "Error submitting OsmosisClPool ICQ, error '%s'", err.Error()))
+		return err
+	}
+
+	if err := k.SetTokenPriceQueryInProgress(ctx, tokenPrice, true); err != nil {
+		k.Logger(ctx).Error(utils.LogWithPriceToken(tokenPrice, "Error updating queryInProgress=true, error '%s'", err.Error()))
+		return err
+	}
+
+	return nil
+}
+
+var queryIdRegex = regexp.MustCompile(`^(.+)\|(.+)\|(\d+)\|(\d+)$`)
+
+// Helper function to parse the ID string
+func parseQueryID(id string) (baseDenom, quoteDenom, poolId, blockHeight string, ok bool) {
+	matches := queryIdRegex.FindStringSubmatch(id)
+	if len(matches) != 5 {
+		return "", "", "", "", false
+	}
+	return matches[1], matches[2], matches[3], matches[4], true
+}
+
+func OsmosisClPoolCallback(k Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
+	// TODO is this check even needed?
+	if query.Id != query.CallbackId {
+		return fmt.Errorf("query.Id ('%s') != query.CallbackId ('%s')", query.Id, query.CallbackId)
+	}
+
+	baseDenom, quoteDenom, poolId, _, ok := parseQueryID(query.Id)
+	if !ok {
+		return fmt.Errorf("Error parsing baseDenom and quoteDenom from queryId '%s'", query.Id)
+	}
```

Comment:
I don't think you really need this - this type of thing is really what the CallbackData is for

I think there are two possible options:
1. Store the token price struct as the callback data (as you have it here), or
2. Use a new callback data struct that only has the keys of the TokenPrice struct, then here, deserialize callback data to get keys and then look up the TokenPrice from those keys

Generally, I think case 2 usually makes more sense so that you grab the latest TokenPrice from the store at callback time, instead of using one that may be a little stale. But I think either is valid. 

Main thing you want to avoid is accidentally writing a stale TokenPrice in the callback that overrides any changes that were done since the query was triggered (e.g. the TokenPrice is removed while a query is in flight and then the callback re-adds it)

Reply:
hear hear my icq king

Reply:
60b2cf0e35390bd02b1bde8ae9d76db2685015f9